src/QsCompiler/CompilerOptimizations/OptimizingTransformations/CallableInlining.fs

```diff
@@ -122,42 +129,87 @@ type internal CallableInliner(callables) =
-    /// Returns None if the expression cannot be safely inlined.
-    let safeInline expr =
+    /// The random generator to give unique variable names
+    let random = System.Random()
+
+    let tryInline expr =
         maybe {
             let! ii = tryGetInliningInfo callables expr
 
-            do! check (countReturnStatements ii.body = 0)
-            let! current = currentCallable
-            do! check (cannotReachCallable callables ii.body current.FullName)
+            let! currentCallable = currentCallable
+            let! renamer = renamer
+            renamer.clearStack()
+            do! check (cannotReachCallable callables ii.body currentCallable.FullName)
             do! check (ii.functors.controlled < 2)
             // TODO - support multiple Controlled functors
             do! check (cannotReachCallable callables ii.body ii.callable.FullName || isLiteral callables ii.arg)
 
             let newBinding = QsBinding.New ImmutableBinding (toSymbolTuple ii.specArgs, ii.arg)
-            let newStatements = ii.body.Statements.Insert (0, newBinding |> QsVariableDeclaration |> wrapStmt)
-            return {ii.body with Statements = newStatements} |> QsScopeStatement.New |> QsScopeStatement
+            let newStatements =
+                ii.body.Statements.Insert (0, newBinding |> QsVariableDeclaration |> wrapStmt)
+                |> Seq.map renamer.Scope.onStatement
+                |> Seq.map (fun s -> s.Statement)
+                |> ImmutableArray.CreateRange
+            return ii, newStatements
         }
 
+    /// Inline an expression representing a callable with no return statements.
+    /// Returns None if the expression cannot be safely inlined.
+    let safeInline expr =
+        maybe {
+            let! ii, newStatements = tryInline expr
+            do! check (countReturnStatements ii.body = 0)
+            return newStatements
+        }
 
-    override this.onCallableImplementation c =
-        let prev = currentCallable
-        currentCallable <- Some c
-        let result = base.onCallableImplementation c
-        currentCallable <- prev
-        result
-
-    override syntaxTree.Scope = { new ScopeTransformation() with
+    /// Inline an expression representing a callable with exactly one return statement.
+    /// Returns None if the expression cannot be safely inlined.
+    let safeInlineReturn expr =
+        maybe {
+            let! ii, newStatements = tryInline expr
 
-        override scope.StatementKind = { new StatementKindTransformation() with
-            override stmtKind.ExpressionTransformation x = scope.Expression.Transform x
-            override stmtKind.LocationTransformation x = scope.onLocation x
-            override stmtKind.ScopeTransformation x = scope.Transform x
-            override stmtKind.TypeTransformation x = scope.Expression.Type.Transform x
+            do! check (countReturnStatements ii.body = 1)
+            let lastStatement = newStatements.[newStatements.Length-1]
```

Comment:
It would be good to document that the assumption is that dead code elimination is done beforehand.